Docs/Proposals/0001-jsonencoder-key-strategy.md

```diff
@@ -0,0 +1,152 @@
+        ///
+        /// Capital characters are determined by testing membership in `CharacterSet.uppercaseLetters` and `CharacterSet.lowercaseLetters` (Unicode General Categories Lu and Lt).
+        /// The conversion to lower case uses `Locale.system`, also known as the ICU "root" locale. This means the result is consistent regardless of the current user's locale and language preferences.
+        ///
+        /// Converting from camel case to snake case:
+        /// 1. Splits words at the boundary of lower-case to upper-case
+        /// 2. Inserts `_` between words
+        /// 3. Lowercases the entire string
+        /// 4. Preserves starting and ending `_`.
+        ///
+        /// For example, `oneTwoThree` becomes `one_two_three`. `_oneTwoThree_` becomes `_one_two_three_`.
+        ///
+        /// - Note: Using a key encoding strategy has a nominal performance cost, as each string key has to be converted.
+        case convertToSnakeCase
+        
+        /// Provide a custom conversion to the key in the encoded JSON from the keys specified by the encoded types.
+        /// The full path to the current encoding position is provided for context (in case you need to locate this key within the payload). The returned key is used in place of the last component in the coding path before encoding.
+        case custom(([CodingKey]) -> CodingKey)
+    }
+    
+    
+    /// The strategy to use for encoding keys. Defaults to `.useDefaultKeys`.
+    open var keyEncodingStrategy: KeyEncodingStrategy = .useDefaultKeys
+}
+```
+
+## Detailed design
+
+The strategy enum allows developers to pick from common actions of converting to and from `snake_case` to the Swift-standard `camelCase`. The implementation is intentionally simple, because we want to make the rules predictable.
+
+Converting from snake case to camel case:
+
+1. Capitalizes the word starting after each `_`
+2. Removes all `_`
+3. Preserves starting and ending `_` (as these are often used to indicate private variables or other metadata).
+
+For example, `one_two_three` becomes `oneTwoThree`. `_one_two_three_` becomes `_oneTwoThree_`.
+
+Converting from camel case to snake case:
+
+1. Splits words at the boundary of lower-case to upper-case
+2. Inserts `_` between words
+3. Lowercases the entire string
+4. Preserves starting and ending `_`.
+
+For example, `oneTwoThree` becomes `one_two_three`. `_oneTwoThree_` becomes `_one_two_three_`.
+
+We also provide a `custom` action for both encoding and decoding to allow for maximum flexibility if the built-in options are not sufficient.
+
+## Example
+
+Given this JSON:
+
+```
+{ "hello_world" : 3, "goodbye_cruel_world" : 10, "key" : 42 }
+```
+
+Previously, you would customize your `Decodable` type with custom keys, like this:
+
+```swift
+struct Thing : Decodable {
+
+    let helloWorld : Int
+    let goodbyeCruelWorld: Int
+    let key: Int
+
+    private enum CodingKeys : CodingKey {
```

Comment:
This should have a raw type of `String`, I think?

Reply:
Yup